### PR TITLE
lang/perl: Upgrade to Perl 5.24.1

### PR DESCRIPTION
**Pull request by michailf**
Signed-off-by: Michail Frolov <frolovm@pobox.com>

Maintainer: @pprindeville
Compile tested: x86_64, compiled using: `make package/feeds/packages/perl/compile`
Run tested: During compile the host-compiled version is used by the Perl make.

### Description:

Upgraded to Perl 5.24.1

### CODE REVIEW

File: lang/perl/patches/310-recompile-hack.patch [hunk 1]
```diff
@@ -5,12 +5,12 @@ patch has been reported to work around that issue.
...
 +++ b/Makefile.SH
-@@ -818,7 +818,7 @@ $(LIBPERL): $& $(obj) $(DYNALOADER) $(LI
+@@ -867,7 +867,7 @@ $(LIBPERL): $& $(perllib_objs) $(DYNALOA
  	true)
  		$spitshell >>$Makefile <<'!NO!SUBS!'
  	rm -f $@
--	$(LD) -o $@ $(SHRPLDFLAGS) $(obj) $(DYNALOADER) $(libs)
-+	$(CC) -o $@ $(SHRPLDFLAGS) $(obj) $(DYNALOADER) $(libs)
+-	$(LD) -o $@ $(SHRPLDFLAGS) $(perllib_objs) $(DYNALOADER) $(libs)
++	$(CC) -o $@ $(SHRPLDFLAGS) $(perllib_objs) $(DYNALOADER) $(libs)
```
Comment (pprindeville) [inline lang/perl/patches/310-recompile-hack.patch]: Why do we need to link with `$(CC)` and not `$(LD)`?  Can't we just change our definition of `LD=...` to be the right thing so this patch isn't needed?

Comment (michailf) [inline lang/perl/patches/310-recompile-hack.patch]: My answer is: I don't know.

The patch was made with the following comment:

> Apparently, recompiling/relinking fails under some circumstances. This
> patch has been reported to work around that issue.
> 
> Signed-off-by: Marcel Denia <naoir@gmx.net>

I could not find Marcel Denia on github so I doubt we will get any answers.

The problem is that I do not know what "circumstances" he was talking about, so I cannot test this.

I would imagine that changing LD= would affect the compile on a more global scale than the patch.

We can try removing the patch completely and see if the problem comes back.

Comment (pprindeville) [inline lang/perl/patches/310-recompile-hack.patch]: > We can try removing the patch completely and see if the problem comes back.

Let's start with that.

Also, can you do (logged) builds with -j1 V=s before and after the change (and then diff them) so we can compare what effectively is different?

Comment (michailf) [inline lang/perl/patches/310-recompile-hack.patch]: Seems like there is no difference.  I ran with and without the 310 patch after running

`make package/feeds/packages/perl/clean`

[lang_perl_compile_patch310.txt](https://github.com/openwrt/packages/files/988399/lang_perl_compile_patch310.txt)
[lang_perl_compile_NOpatch310.txt](https://github.com/openwrt/packages/files/988398/lang_perl_compile_NOpatch310.txt)

Comment (pprindeville) [inline lang/perl/patches/310-recompile-hack.patch]: I concur that there seems to be no significant difference.

**Comment by diizzyy** [comment]
Being nice netizens can we update the URL list and offload the main site?
```
https://cpan.uib.no/src/5.0
https://mirror.uic.edu/CPAN/src/5.0
https://mirror.transip.net/CPAN/src/5.0
https://mirrors.sonic.net/cpan/src/5.0
https://www.cpan.org/src/5.0
```

**Comment by michailf** [comment]
@diizzyy But why these 5 sites out of 427 available?  Would be nice if there was an ability to pick closest regional site.

File: lang/perl/patches/310-recompile-hack.patch (hunk 1 above)
Comment (michailf) [inline lang/perl/patches/310-recompile-hack.patch]: Done

**Comment by michailf** [comment]
metacpan.org seems to resolve to nearest location.  How about we use the following list?

https://cpan.metacpan.org/src/5.0
https://mirrors.rit.edu/CPAN/src/5.0
https://mirrors.sonic.net/cpan/src/5.0
https://mirror.uic.edu/CPAN/src/5.0
https://cpan.uib.no/src/5.0
https://mirror.transip.net/CPAN/src/5.0
https://mirror.downloadvn.com/cpan/src/5.0
https://www.cpan.org/src/5.0

**Comment by diizzyy** [comment]
metacpan is hosted using the fastly CDN (wherever that is).
I would however recommend you not to only place NA hosts first.

My list is based on overall peering where most users are (EU/NA in general) and availability hence the mix. This is already what we do for kernel, gnu, savannah, apache, gnome packages (domains) in since a year back now and most seem happy.

**Comment by hnyman** [comment]
> cpan.metacpan.org seems to resolve to nearest location. 

> metacpan is hosted using the fastly CDN (wherever that is).

To me it sounds good that a global CDN site is used first.
As there is a CDN site available, use that and then 2-3 other sites scattered around the globe.

The length of the list not make much difference, as it just provides fallback sites in case there is connectivity problems with the first site.  (The list is evaluated in order)

**Comment by michailf** [comment]
Does the following work?

`https://cpan.metacpan.org/src/5.0`
`https://cpan.uib.no/src/5.0` _# Norway_
`https://mirrors.rit.edu/CPAN/src/5.0` _# U.S. east coast_
`https://mirror.transip.net/CPAN/src/5.0` _# Netherlands_
`https://mirrors.sonic.net/cpan/src/5.0` _# U.S. west coast_
`https://mirror.downloadvn.com/cpan/src/5.0` _# Viet Nam_
`https://www.cpan.org/src/5.0`

metacpan.org would normally resolve to a local region in the first place. (_Hong Kong, Japan, Singapore, UAE, France, Germany, Netherlands, Spain, Sweden, England, Canada, USA, Australia, New Zealand, Brazil_)

I tried to find sites that support https as that seems to be what you were going for.

**Comment by michailf** [comment]
Or simply this:

```
https://cpan.metacpan.org/src/5.0
https://www.cpan.org/src/5.0
```

**Comment by diizzyy** [comment]
Use the first one and drop vietnam :-)

**Comment by pprindeville** [comment]
Thanks!

**Pull request merged**